### DISCUSSION TITLE
Return or receive another module's type

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12159] [Ideas]
## Problem

A module cannot return another module's type, or receive it as argument. This makes it harder to break down larger projects into several modules.

## Solution

I'm not sure what the solution is.

- On the one hand, the lack of this feature is becoming quite painful
- On the other hand, I know that this has been considered, and the conclusion was: it would create intractable "dependency hell" problems.

I wonder if the dependency hell problems could be solved with name mangling?

For example:

- Module `App` imports another module `Go`
- `App.builder()` returns a `Go.Builder`
- The assumption has been, I think, that the resulting binding would be `func (app App) Builder() *dagger.GoBuilder`. Which could cause dependency hell
- But, perhaps instead the signature could be: `func (app App) Builder() *dagger.AppGoBuilder`. In other words, `GoBuilder` as imported by `App` would get its own namespace. I think this might make the dependency hell problem go away.

Does this have legs @sipsma @vito @jedevc @helderco ?

**Comment by jedevc** [comment]
The original logic for this is illustrated in https://github.com/dagger/dagger/pull/6102#issuecomment-1813209934.

There was some discussion recently about this in https://discord.com/channels/707636530424053791/1285289216556925002/1285592864448974999.

--- 

> * In other words, `GoBuilder` as imported by `App` would get its own namespace. I think this might make the dependency hell problem go away.

I don't think this works - if you return a `AppGoBuilder` from a module that's not `App`, then this type just propagates all through the dependency tree. The problem that can occur is this:
- B returns a type from A version 1
- C returns a type from A version 2
- D calls both B and C - these types are potentially different, but under the above scheme, they'd be the same name.

**Comment by shykes** [comment]
If they are called `FooAppGoBuilder` and `BarAppGoBuilder` doesn't that solve the propagation problem?

**Comment by jedevc** [comment]
I think this potentially solves this, but at the cost of making it very very difficult to use - the `App` name now has to propagate through the entire system - for deeply nested modules, this produces names that would be fairly unreadable.

It also means we'll struggle with caching - with different names in the graphql schema depending on the caller, different callers won't cache. Maybe we could inject the version into the name?

---

Actually! We potentially have a neat solution I think now, we can use dagql views (that power module compat). Just to throw another option into the ring.

To illustrate, the following scenario:
- B returns a type from A version 1 - it can't directly construct the type, so it's had to return an ID for it (which corresponds to a call to A). This ID can be annotated with a view - which is "version 1".
- C returns a type from A version - now this ID is annotated with a *different* view, "version 2".
- A can access both B and C easily - however, when we need to work with these objects, we can use these views to dispatch the calls to the right version of the module - we never send the data for the encoded objects to the wrong module.

However, there's one disadvantage to this - it requires that the fields of the type that A accesses must all exist in B and C. This is kinda fiddly to do, and we can really only do it at runtime, so might cause some weird errors and behaviors.

---

The other idea I had in the discord thread is if we could work out a scenario where we could *always* drop this restriction, where we work out that diamond dependencies are impossible.

One potential option would be for an "internal" module - in this case, an internal module is only importable via a local path (and not a remote git source). This is pretty much what @sipsma has called "module extension" - in this case, it's safe to do, because essentially we're *re-exporting* the types.

**Comment by sipsma** [comment]
I ruled this idea out when we first started thinking about this problem for a few reasons:

One is what Justin mentioned about "for deeply nested modules, this produces names that would be fairly unreadable"; this approach might be okay for dependencies 1 level deep, but it really breaks down in terms of comprehensibility after that. 

Imagine a dependency graph of modules that looks like:
```
Bar 
     \
      Foo - Blah - Qaz - Common 
     /
Baz
```

If your module has a dep on both `Bar` and `Baz` you can end up with types named like `BarFooBlahQazCommon` and `BazFooBlahQazCommon`.

On top of that, at least in Go, even those two types *might* really be the same, they could *never* be the same in terms of the compiler, so you'd need to do explicit conversions between them even in the case where they end up being the same.
* In Go, one typical way of dealing with this sort of problem is an `interface`, which is where the idea of supporting interfaces across languages came about.

---

Generally speaking, I think that the path forward here should be:
1. Continue to flesh out interfaces + smooth out DX warts (e.g. https://github.com/dagger/dagger/issues/6366)
2. Explore what @jedevc and I were talking about in that discord thread that he summarized here well
